api/v1alpha1/backend_types.go

```diff
@@ -0,0 +1,157 @@
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=envoy-gateway,shortName=be
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=`.status.conditions[?(@.type=="Accepted")].reason`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+// +notImplementedHide
+//
+// Backend allows the user to configure the behavior of the connection
+// between the Envoy Proxy listener and the backend service.
+type Backend struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec defines the desired state of Backend.
+	Spec BackendSpec `json:"spec"`
+
+	// status defines the current status of Backend.
+	Status BackendStatus `json:"status,omitempty"`
+}
+
+// BackendAddress describes are backend address, which is can be either a TCP/UDP socket or a Unix Domain Socket
+// corresponding to Envoy's Address: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-address
+//
+// +kubebuilder:validation:XValidation:rule="(has(self.socketAddress) || has(self.unixDomainSocketAddress))",message="one of socketAddress or unixDomainSocketAddress must be specified"
+// +kubebuilder:validation:XValidation:rule="(has(self.socketAddress) && !has(self.unixDomainSocketAddress)) || (!has(self.socketAddress) && has(self.unixDomainSocketAddress))",message="only one of socketAddress or unixDomainSocketAddress can be specified"
+// +kubebuilder:validation:XValidation:rule="((has(self.socketAddress) && (self.type == 'FQDN' || self.type == 'IPv4')) || has(self.unixDomainSocketAddress) && self.type == 'UDS')",message="if type is FQDN or IPv4, socketAddress must be set; if type is UDS, unixDomainSocketAddress must be set"
+// +notImplementedHide
+type BackendAddress struct {
+	// Type is the the type name of the backend address: FQDN, UDS, IPv4
+	Type AddressType `json:"type"`
+
+	// SocketAddress defines a FQDN or IPv4 address
+	//
+	// +optional
+	SocketAddress *SocketAddress `json:"socketAddress,omitempty"`
+
+	// UnixDomainSocketAddress defines the unix domain socket path
+	//
+	// +optional
+	UnixDomainSocketAddress *UnixDomainSocketAddress `json:"unixDomainSocketAddress,omitempty"`
+}
+
+// SocketAddress describes TCP/UDP socket address, corresponding to Envoy's SocketAddress
+// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-socketaddress
+// +notImplementedHide
+type SocketAddress struct {
+	// Address defines to the FQDN or IP address of the backend service.
+	Address string `json:"address"`
+
+	// Port defines to the port of of the backend service.
+	//
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65535
+	Port int32 `json:"port"`
+
+	// Protocol defines to the the transport protocol to use for communication with the backend.
+	//
+	// +optional
+	Protocol *ProtocolType `json:"protocol,omitempty"`
```

Comment:
can we do w/o this field in the init PR and rely on info from the `xRoute` for the protocol ?

Reply:
Would this result in protocol mismatching if the protocol of `xRoute` is not the same as the `Backend`?

Reply:
that depends on the backend server type, so app dev should be using the right Route type imo

Reply:
We can add validation to check if there's a protocol mismatching if the Backend API also has a Protocol field and reflect it to the xRoute status. Otherwise, the mismatching leaks to runtime, making it harder for debugging.

Reply:
yeah my vote would be to capture this as a issue and tackle once the need arises